### PR TITLE
Feature/gh 220 refactors slurm jobs

### LINKED ISSUE #220: Refactor Slurm Jobs
**As:** Yorc dev team
**I want to:** Refactor how Slurm Jobs are modeled
**So that:** It could be more user-friendly

* [x] **AC1:** Distinction between srun and sbatch jobs does not make sens anymore and make it more complex
* [x] **AC2:** We should be able to run a job only using command line args without providing an artifact
* [ ] **AC3:** Ideally if singularity jobs could be factorized with other jobs TOSCA desc it would be nice. Maybe only the implementation type should be enough to differentiate job types
* [x] **AC4:** Simplify the implementation in YORC
* [x] **AC4:** Retrieve real Slurm job status once runned

### PR DESCRIPTION
**Pull request by stebenoist**
# Pull Request description
Refactor Slurm Jobs
As: Yorc dev team
I want to: Refactor how Slurm Jobs are modeled
So that: It could be more user-friendly

AC1: Distinction between srun and sbatch jobs does not make sens anymore and make it more complex
AC2: We should be able to run a job only using command line args without providing an artifact
AC3: Ideally if singularity jobs could be factorized with other jobs TOSCA desc it would be nice. Maybe only the implementation type should be enough to differentiate job types
AC4: Simplify the implementation in YORC
AC4: Retrieve real Slurm job status once runned

## Description of the change
All ACs OK except AC3: Singularity Job still exists with specific properties related to Singularity execution
### What I did
- Job are runned asynchronously so only Slurm sbatch is supported.
- A command can be executed as a job (no need to provide any artifact) with **command** and **args** properties
- Environment variables can be explicitly added with **env_vars** property.
- Add **working_directory** Tosca prop to let users define the working directory for Slurm sbatch execution and check outputs, stderr, stdout relative to this directory
- **keep_job_remote_artifacts** Slurm infrastructure property allows not to remove artifacts when a job is done.
- Slurm scontrol command is used for monitoring job (allows to get job status, stdout/stderr)

### How to verify it

- Create yorc orchestrator
- Create Slurm location
- Add Job and SinglarityJob on-demand resources in Slurm location
- Add https://github.com/ystia/yorc-a4c-plugin/tree/feature/gh-220-refactors-slurm-jobs/tosca-samples to 
  catalog
- Check and deploy the following topologies:

 **For Slurm job**
 - command
 - mpi
 - mpi2
 - failure

**For Singularity Slurm job:**

- hello
- lolcow
- bash_loop
- compute_pi
- mnist

### Description for the changelog
* Refactor Slurm jobs ([GH-220](https://github.com/ystia/yorc/issues/220))

## Applicable Issues
fixes #220

### CODE REVIEW

File: data/tosca/yorc-slurm-types.yml [hunk 1]
```diff
@@ -97,22 +97,32 @@ node_types:
           Set a limit on the total run time of the job allocation.
           Time formats include "minutes", "minutes:seconds", "hours:minutes:seconds", "days-hours", "days-hours:minutes" and "days-hours:minutes:seconds"
         required: false
-      batch:
-        type: boolean
-        description: True if the job is run on batch mode.
-        required: false
-        default: true
-      extra_options:
+      extra_job_options:
```
Comment (loicalbertin) [inline data/tosca/yorc-slurm-types.yml]: Maybe it's time to group properties functionally:

- using data types if possible
- or at least in giving them in a coherent order (in a4c plugin)

File: prov/slurm/execution.go [hunk 2]
```diff
@@ -151,29 +145,37 @@ func (e *executionCommon) execute(ctx context.Context) error {
 	// Fill log optional fields for log registration
 	switch strings.ToLower(e.operation.Name) {
 	case strings.ToLower(tosca.RunnableSubmitOperationName):
-		log.Printf("Running the job: %s", e.operation.Name)
-
+		log.Printf("Submit the job: %s", e.operation.Name)
```
Comment (loicalbertin) [inline prov/slurm/execution.go]: Should be debug

File: data/tosca/yorc-slurm-types.yml [hunk 3]
```diff
@@ -69,50 +117,30 @@ node_types:
...
-        default: 1
-      nodes:
-        description: Number of nodes allocated to the job.
-        type: integer
-        required: false
-        default: 1
-      cpus_per_task:
-        description: Number of cpus allocated per task.
-        type: integer
+      job_properties:
```
Comment (loicalbertin) [inline data/tosca/yorc-slurm-types.yml]:
```suggestion
      slurm_options:
```
As I understand it you grouped properties that are translated into Slurm options (I may be wrong) so I suggest to rename this property

Comment (stebenoist) [reply data/tosca/yorc-slurm-types.yml]: right

**Pull request merged**